python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/shape_lib_gen.py

```diff
@@ -1122,6 +1122,58 @@ def aten〇index〇Tensor(self: List[int], indices: List[Optional[List[int]]]) -
 
+@check_shape_function([
+    Invocation(TensorOfShape(2), [LongTensorOfShape(4)]), # Basic case.
+    Invocation(TensorOfShape(2, 3), [LongTensorOfShape(4), LongTensorOfShape(4)]), # More dimensions.
+    Invocation(TensorOfShape(2, 3), [LongTensorOfShape(4), LongTensorOfShape(6, 4)]), # Multidimensional index tensor along a dimension.
+    Invocation(TensorOfShape(2, 3), [LongTensorOfShape(4), None]), # Explicit None value.
+    Invocation(TensorOfShape(2, 3, 4, 5), [None, LongTensorOfShape(4), LongTensorOfShape(4)]), # Indexing tensors on consecutive dimensions.
+    Invocation(TensorOfShape(2, 3, 4, 5), [None, LongTensorOfShape(4), None, LongTensorOfShape(4)]), # Indexing tensors on non-consecutive dimensions.
+    Invocation(TensorOfShape(2, 3, 4, 5), [LongTensorOfShape(4, 2), None, LongTensorOfShape(2)]), # Indexing tensors on non-consecutive dimensions.
+    Invocation(TensorOfShape(2, 3), [LongTensorOfShape(4, 5, 6), LongTensorOfShape(1, 5, 1)]), # Broadcasting of index tensors.
+    Invocation(TensorOfShape(2, 3), [LongTensorOfShape(4)]), # Fewer index tensors than dimensions.
+    ErrorInvocation(TensorOfShape(2, 3), [LongTensorOfShape(4), LongTensorOfShape(4), LongTensorOfShape(4)]), # More index tensors than dimensions.
+])
+def aten〇index〇Tensor_hacked_twin(self: List[int], indices: List[List[int]]) -> List[int]:
+    assert len(indices) <= len(self), "More indices than dimensions to index"
+    broadcasted_shape: List[int] = []
+    unused_dim_sizes: List[int] = []
+    for i in range(len(self)):
+        if i >= len(indices):
+            unused_dim_sizes.append(self[i])
+        else:
+            index_tensor_shape = indices[i]
+            if index_tensor_shape is not None:
+                broadcasted_shape = upstream_shape_functions.broadcast(broadcasted_shape, index_tensor_shape)
+            else:
+                unused_dim_sizes.append(self[i])
+
+    if len(unused_dim_sizes) == 0:
+        return broadcasted_shape
+
+    first_index_tensor_location = -1
+    index_tensors_are_together = True
+    for e, index_tensor_shape in enumerate(indices):
+        if index_tensor_shape is not None:
+            if first_index_tensor_location == -1:
+                first_index_tensor_location = e
+            elif e - first_index_tensor_location != 1:
+                index_tensors_are_together = False
+
+    if not index_tensors_are_together:
+        return broadcasted_shape + unused_dim_sizes
+
+    # If index tensors are all in consecutive dimensions, the broadcasted
+    # shape is inserted in the location of the consecutive index tensors.
+    result_shape: List[int] = []
+    for i in range(first_index_tensor_location):
+        result_shape.append(unused_dim_sizes[i])
+    for broadcasted_size in broadcasted_shape:
+        result_shape.append(broadcasted_size)
+    for i in range(first_index_tensor_location, len(unused_dim_sizes)):
+        result_shape.append(unused_dim_sizes[i])
+    return result_shape
+
```

Comment:
This is the exact same code used for `aten.index.Tensor`. Please refactor the code into a helper function to avoid repetition.

Reply:
Hi @ramiro050, TBH I don't see a way of using a common helper for both the ops, since the type of indices for `index.tensor` is `List[Optional[List[int]]]`, and for `index.tensor_hacked_twin` is `List[List[int]]`.

Reply:
`List[List[int]]` is a subtype of `List[Optional[List[int]]]`, so if the helper function takes as an argument a `List[Optional[List[int]]]`, it should work for both types

Reply:
I have tried this, but it's not working.
```
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/vivek/work/02_07/vivekkhandelwal1-torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir/dialects/torch/importer/jit_ir/build_tools/shape_lib_gen.py", line 1356, in <module>
    main(_create_argparse().parse_args())
  File "/home/vivek/work/02_07/vivekkhandelwal1-torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir/dialects/torch/importer/jit_ir/build_tools/shape_lib_gen.py", line 1284, in main
    torch.jit.script(v)
  File "/home/vivek/work/02_07/mlir_venv/lib/python3.8/site-packages/torch/jit/_script.py", line 1343, in script
    fn = torch._C._jit_script_compile(
RuntimeError: 

index_Tensor_Like(int[] self, int[]?[] indices) -> int[]:
Expected a value of type 'List[Optional[List[int]]]' for argument 'indices' but instead found type 'List[List[int]]'.
:
  File "/home/vivek/work/02_07/vivekkhandelwal1-torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir/dialects/torch/importer/jit_ir/build_tools/shape_lib_gen.py", line 1152
])
def aten〇index〇Tensor_hacked_twin(self: List[int], indices: List[List[int]]) -> List[int]:
    index_Tensor_Like(self, indices)
    ~~~~~~~~~~~~~~~~~ <--- HERE
    assert len(indices) <= len(self), "More indices than dimensions to index"
    broadcasted_shape: List[int] = []
```

Reply:
Hm, that's strange. I think you should be able to get around this by doing something like

```python
def aten〇index〇Tensor_hacked_twin(self: List[int], indices: List[List[int]]) -> List[int]:
    optional_indices: List[Optional[List[int]]] = [x for x in indices]
```

to turn the `indices` to the right type